### ISSUE TITLE
implement NPE testing framework

### ISSUE THREAD

**Issue by changhoonhahn** [galactic-ai/haloflow#3]
implement validation and testing framework for NPEs

**Comment by changhoonhahn** [comment]
NPE validation script implemented [6010a83](https://github.com/galactic-ai/haloflow/commit/6010a83bc580e75600e298b0c576fe060ea422a5). Validation includes simulation-based calibration
![Image](https://github.com/user-attachments/assets/0a1ccb09-f354-4a0c-b34b-13530d4e665b)
and TARP
![Image](https://github.com/user-attachments/assets/9560a740-6647-4a6a-b7ad-427721a5318e). 

A quick check suggests we need to train a large ensemble of NPEs. Deployed.

**Comment by changhoonhahn** [comment]
Further testing/validation implemented in [394b707](https://github.com/galactic-ai/haloflow/commit/394b707f1445f93a0af2e87f41aa0d540db3a708)

Here's a figure illustrating the performance of the trained NPEs: 
![image](https://github.com/user-attachments/assets/960a494d-9859-4ad1-93f9-8b051a76a0c5)

Takeaways: 
- The prior seems to be affecting some of the posteriors
- Further training is necessary for TNG50
- Something is going completely wrong for TNG100

**Comment by changhoonhahn** [comment]
Preliminary NPE validation ([f6b6fe3](https://github.com/galactic-ai/haloflow/commit/f6b6fe33568c7be398635ce4429d1ccde1569305)) with 
- updated halo masses
- TNG_ALL with mass and magnitude corrections

![image](https://github.com/user-attachments/assets/3fce7b43-a522-4d14-b91b-fd44b213be28)

Looks reasonable but further NPE training is necessary.

**Comment by Nikhil0504** [comment]
Some more training done: see 069f2dcf8ea8d44548e0c24d41ad5296ac653579

![Image](https://github.com/user-attachments/assets/4bcddda1-47d1-463a-b8d1-5f9448ffcff9)

Unsure if we can get better than this. A few more hyperparameter sweeps are ongoing but it's very computationally expensive.  

@changhoonhahn do you think this is enough training and testing to move on to #10 ?

**Comment by changhoonhahn** [comment]
@Nikhil0504, by eye the NPEs look quite reasonable. 

Before you stop the hyperparameter sweeps, do you have the coverage tests of _trained on X, tested on X_? The rank distribution should look close to uniform and the DRP coverage test should look close to diagonal.

**Comment by Nikhil0504** [comment]
@changhoonhahn  
Yes, I do have those.  Here are the plots. This is only for `mags` observables.

## Eagle 100
![Image](https://github.com/user-attachments/assets/80f1c84b-9a1f-4d00-92c7-0de77052f9c6)
![Image](https://github.com/user-attachments/assets/aa77dfde-f4fd-4e89-b48c-994e3cfd7405)

## Simba 100
![Image](https://github.com/user-attachments/assets/95a0f998-63a4-4918-a3c6-e8fae73d307b)
![Image](https://github.com/user-attachments/assets/076b98ac-1a38-49c7-85c3-7e25bd8a508a)

## TNG 100
![Image](https://github.com/user-attachments/assets/8c629007-757b-429a-9555-02584ab9cb80)
![Image](https://github.com/user-attachments/assets/2ecd947b-c2de-441d-ada7-65cd97697de4)

## TNG ALL
![Image](https://github.com/user-attachments/assets/ad81522a-5082-4ba1-bedc-73ac98bb368e)
![Image](https://github.com/user-attachments/assets/087567f3-74c7-43f8-94f1-d27824e659d6)

## TNG 50
![Image](https://github.com/user-attachments/assets/6cf2d8cd-d68d-4456-bccb-838cc37ebd00)
![Image](https://github.com/user-attachments/assets/7633b696-273c-482e-adcf-9095efc49318)

**Comment by changhoonhahn** [comment]
These look good. I'd say this is enough training.

Let's make a DRP coverage test plot with all the NPEs in a single panel with different colors for the paper then close this out.

**Comment by Nikhil0504** [comment]
Here are the coverage test plots. Implemented in f3b18ab

For `mags`
![Image](https://github.com/user-attachments/assets/d0189a23-adae-4b0d-8e9f-152228569fb3)

For `mags_morph`
![Image](https://github.com/user-attachments/assets/e8bb2dbd-cdf0-4b7a-8fa6-53e19287a35c)
